socket_mode.go

```diff
@@ -0,0 +1,92 @@
+package slack
+
+import (
+	"context"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+const (
+	socketModeEventTypeHello      = "hello"
+)
+
+// StartSocketMode calls the "rtm.start" endpoint and returns the provided URL and the full Info block.
+//
+// To have a fully managed Websocket connection, use `NewRTM`, and call `ManageConnection()` on it.
+func (api *Client) StartSocketMode() (info *SocketModeConnection, websocketURL string, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), websocketDefaultTimeout)
+	defer cancel()
+
+	return api.StartSocketModeContext(ctx)
+}
+
+type openResponseFull struct {
+	SlackResponse
+	SocketModeConnection
+}
+
+// StartSocketModeContext calls the "apps.connections.open" endpoint and returns the provided URL and the full Info block with a custom context.
+//
+// To have a fully managed Websocket connection, use `NewRTM`, and call `ManageConnection()` on it.
+func (api *Client) StartSocketModeContext(ctx context.Context) (info *SocketModeConnection, websocketURL string, err error) {
+	response := &openResponseFull{}
+	err = api.postMethod(ctx, "apps.connections.open", url.Values{"token": {api.token}}, response)
```

Comment:
The token here needs to be an app-level token.

Reply:
@mumoshu app-level tokens were PRed in a while back, so `api.appLevelToken` should work. The user sets it by passing `OptionAppLevelToken` to the client definition.

Reply:
@seratch @cjdenio Thanks to your comments I updated it to use `appLevelToken` for `apps.connection.open` and it's now working 👍